bot/Engine/BotController.py

```diff
@@ -0,0 +1,385 @@
+        if order is not None:        
+            new_order_model.matched_order_id = order.id
+            order.is_closed = True
+            order.matched_order_id = new_order_model.id
+            pair.active = False
+            pair.current_order_id = new_order_model.id
+
+        if order is None:
+            self.bot.current_balance = self.bot.current_balance - Decimal(self.bot.starting_balance) * Decimal(self.bot.entry_settings.initial_entry_allocation) / 100
+            pair.active = False
+            pair.current_order_id = new_order_model.id    
+
+    def computeQuantity(self, order):
+        """ compute exit quantity so that also partially filled orders can be handled."""
+        if order.side == 'BUY':
+            exit_quantity = order.executed_quantity
+        elif order.side == 'SELL':
+            exit_quantity = Decimal(order.original_quantity) - Decimal(order.executed_quantity)
+        return exit_quantity
+    
+    def simulateOrderInfo(self, order):
+        """ Used when BotController is in test mode. Simulates order info returned by exchange."""
+        order_status = dict()
+        candlestick_data = self.exchange.getSymbolKlines(order.symbol, self.kline_interval, self.strategy.minimum_period)
+        current_price = candlestick_data.iloc[-1]['close']
+        if order.side == 'BUY':
+            # Market Entry Order
+            if self.bot.entry_settings.signal_distance == 0:
+                order_status['status'] = self.exchange.ORDER_STATUS_FILLED
+                order_status['side'] = order.side
+                order_status['executedQty'] = order.original_quantity
+            # Limit Entry Order
+            elif (Decimal(100) + Decimal(self.bot.entry_settings.signal_distance))/Decimal(100) * Decimal(order.entry_price) <= current_price:
+                order_status['status'] = self.exchange.ORDER_STATUS_FILLED
+                order_status['side'] = order.side
+                order_status['executedQty'] = order.original_quantity
+
+        elif order.order_type == self.exchange.ORDER_TYPE_LIMIT:
+            if current_price >= order.price:
+                order_status['status'] = self.exchange.ORDER_STATUS_FILLED
+                order_status['side'] = order.side
+                order_status['executedQty'] = order.original_quantity
+            else:
+                order_status['status'] = self.exchange.ORDER_STATUS_NEW
+                order_status['side'] = order.side      
+                order_status['executedQty'] = 0
+        elif order.order_type == self.exchange.ORDER_TYPE_MARKET:
+            order_status['status'] = self.exchange.ORDER_STATUS_FILLED
+            order_status['side'] = order.side
+            order_status['executedQty'] = order.original_quantity
+        elif order.order_type == self.exchange.ORDER_TYPE_STOP_LOSS:
+            if ((Decimal(100) - Decimal(self.bot.exit_settings.stop_loss_value))/Decimal(100)) * Decimal(order.entry_price) >= Decimal(current_price):
```

Comment:
is entry_price the limit price? (price we place the limit orders at?)

Reply:
No it is not. Here we can again just use price or entry_price instead of recalculating. It is a mistake on my part.

For entry orders entry_price and price is the same thing. But for exit orders entry_price is the price we entered a trade with. This we need to be able to check if stop-loss or take-profit conditions are met as we otherwise have no way of knowing what price we entered the trade at.